Inclusion of original coroutines test utilities

eder7 opened the pull request:
# Scope

- Update coroutines version
- Added coroutine test dependency
- Deprecate old `testCoroutine` and moved it to `legacy` sub-package
- Added `onTearDown` capability in workflow (was needed on the way)
- Configuration options added
  - `useLegacyCoroutineScope()` to use old-style `testCoroutine` (that needs to be specified in all tests which use the old `testCoroutine` function, that gives time for migration)
  - `autoSetMainCoroutineDispatcher(Boolean)` (default: true) - automatically calls `Dispatchers.setMain()`
  - `autoCancelTestCoroutines(Boolean)` closes open jobs after tests (default: false)
    - I struggle with the decision whether to put this to default: true. Doing that might hide exceptions that go unnoticed, e.g. when a job remains unintentionally active. In that case `runBlockingTest` would usually fail the test in case there are still some coroutines running. But when auto cancel is enabled there would be no such exception. But that's of course an extremely seldom problem. In case, there should be a test for whether the coroutine is still running. In general it would be more convenient to automatically cancel coroutines.
  - The two functions that receive a boolean value are implemented in a way that conflicting configurations lead to an exception, e.g. in steps A you say `true` and in steps B you say `false` for `autoCancelTestCoroutines`. Reasoning: each steps can lay out its intentions. If the intentions differ there is a conflict. You can compare it to dependencies, where you can set a dependency to be `requireMock<Z>`, so a subsequent call of `requireReal<Z>` would fail. Is the approach needed that way or overcomplicating things too much
- `Steps` don't implement `CoroutineScope` anymore (reason: `CoroutineScope` should only be implemented for things that really represent a kind of coroutine scope, but not to represent something that "has" a coroutine scope). Instead there is a new extension function `coroutineScope` which is available on steps _and_ tests. For the legacy version the extension from the `legacy` sub-package has to be used.

# ToDo

- [x] Code cleanup
- [x] Improve naming of `CoroutinesTestContextProvider `
- [x] Add more deprecation annotations 
- [x] Decide whether to use just `coroutineScope` or also `testCoroutineScope`
- [x] Unit tests

xa17d commented:
Thanks for the detailed description :)
Without looking into code details yet, I have some feedback:

> Deprecate old `testCoroutine` → Add new `runBlockingSweetest`

Why not retain the old `testCoroutine` and use the new scope by default, and the old scope if `useLegacyCoroutineScope()` is configured? That would require less code changes for migration and avoid a new method users have to remember.

> `autoCleanupTestCoroutines(Boolean)` closes open jobs after tests (default: false)
> [...] I struggle with the decision whether to put this to default: true.

IMO, having it `false` is the safer option and thus `false` should be the default.

> The two functions that receive a boolean value are implemented in a way that conflicting configurations lead to an exception.
> [...] Is the approach needed that way or overcomplicating things too much❓

Nice edge case thinking! I think it's absolutely necessary. By failing instead of assuming a default you can save someone hours of debugging. We should never write code that can be in a unexpected state due to foreseeable mis-configuration. So please leave it this way 👍.

> `Steps` will soon not be implementing `CoroutineScope` anymore, instead there are new extension function `coroutineScope` and `testCoroutineScope`.

Similar to above, why having both `coroutineScope` and `testCoroutineScope`? There could be only one `coroutineScope` and it could behave differently depending on the configuration with `useLegacyCoroutineScope()`. Coming back to `autoCleanupTestCoroutines`: There could be also a `autoCleanupCoroutineScope`.

Thinking about it again, I cannot remember why `coroutineScope` was necessary, because `runBlockingTest` provides a `CoroutineScope` per receiver 🤔

eder7 commented:
Thanks for the feedback 👍 

> > `Steps` will soon not be implementing `CoroutineScope` anymore, instead there are new extension function `coroutineScope` and `testCoroutineScope`.
> Similar to above, why having both `coroutineScope` and `testCoroutineScope`? There could be only one `coroutineScope` and it could behave differently depending on the configuration with `useLegacyCoroutineScope()`.

Both extensions return the same instance. Usually it's `TestCoroutineScope`, but when it's set to legacy it's a standard `CoroutineScope`. `testCoroutineScope` is just a convenience function so you don't have to cast to `TestCoroutineScope` on your own every time you need it.

> Coming back to `autoCleanupTestCoroutines`: There could be also a `autoCleanupCoroutineScope`.

What is the difference actually? 🤔

> Thinking about it again, I cannot remember why `coroutineScope` was necessary, because `runBlockingTest` provides a `CoroutineScope` per receiver 🤔

Usually you put `runBlockingTest` around a test function. But in sweetest you might need access to the (test) coroutine scope from any place (e.g. a steps class).

eder7 commented:
* Discarded todo: Make context for test coroutine scope configurable (there is not a single use case for that in mySugr's production code)
* New question: I think it's confusing to have `coroutineScope` and `testCoroutineScope` to choose from. Alternative idea: just keep `coroutineScope: CoroutineScope`, but also add `delayController` and `uncaughtExceptionCaptor`. ❓

xa17d commented:
> Both extensions return the same instance. Usually it's TestCoroutineScope, but when it's set to legacy it's a standard CoroutineScope. testCoroutineScope is just a convenience function so you don't have to cast to TestCoroutineScope on your own every time you need it.
> [...]
> New question: I think it's confusing to have coroutineScope and testCoroutineScope to choose from. Alternative idea: just keep coroutineScope: CoroutineScope, but also add delayController and uncaughtExceptionCaptor. ❓

Yes, I think it's confusing. Why not simply making `coroutineScope` a `TestCoroutineScope`? It's a new property, so no legacy code is using it. Thus, when using it, it's not allowed to use it in legacy mode.

A `delayController` is a also a new concept and has a similar impact as having two different scopes IMO. I opt for having one scope which is `TestCoroutineScope` and that's it.

Any additional concept that is anyway provided by Kotlin out-of-the-box is kind of a "vendor-lock-in" for sweetest, which may repel users.

xa17d commented:
> > Coming back to `autoCleanupTestCoroutines`: There could be also a `autoCleanupCoroutineScope`.
>
> What is the difference actually? 🤔

Jobs that are launched with `coroutineScope.launch {...}` cause a test to fail if they are active at the end of the test;
Jobs that are launched with `autoCleanupCoroutineScope.launch {...}` are automatically cancelled at the end of a test without causing a test to fail.

But thinking about it again, just calling `job.cancel()` at the end of a test consumes a similar amount of brain-power as deciding between `coroutineScope` and `autoCleanupCoroutineScope` or configuring an auto-cleanup.

So, I'd opt for keeping it simple and just don't provide an auto-cleanup option.

eder7 commented:
> Why not simply making `coroutineScope` a `TestCoroutineScope`?

That's what I did now 👍 

> no legacy code is using it

That's true for old code, but I removed the `CoroutineScope` interface on `Steps`, so somehow you need to obtain it. So there _is_ legacy code using it.

However, I moved the legacy extensions to a `legacy` sub-package. I hope that helps to clarify things and also outlines the usage better.

> Any additional concept that is anyway provided by Kotlin out-of-the-box is kind of a "vendor-lock-in" for sweetest, which may repel users.

What does that mean in the current case?

xa17d commented on build.gradle:
```diff
@@ -6,7 +6,7 @@ buildscript {
             gradlePlugin: '3.1.3',
             kotlin      : '1.3.10',
-            coroutines  : '1.0.1',
+            coroutines  : '1.3.2',
```
Could be `1.3.3` already :)

xa17d commented on sweetest/src/main/java/com/mysugr/sweetest/framework/build/BaseBuilder.kt:
```diff
@@ -100,6 +107,39 @@ abstract class BaseBuilder<TSelf, TResult : BaseAccessor>(
+     * If there are conflicting calls to this function, an exception is thrown (e.g. in the test `true` is passed, but
+     * in one of the steps `false` is passed).
+     */
+    fun autoSetMainCoroutineDispatcher(value: Boolean) = apply {
```
Just an idea: Since only `testContext` is used here (which is effectively public), this could be an extension function on `BaseBuilder`.
The advantage would be that all the Coroutine Code could reside together, all the Mocking Code could reside together, etc. In the long run, we could extract those parts and could provide _sweetest_ with a _coroutines_-extension, a _mocking_-extension,... At the bottom line: modularity.

eder7 replied:
that's a good point actually. on the other side i planned on ditching the idea to have extensions for these kinds of tasks so people can just control + click members and immediately what the classes are capable of. that would imply e.g. `invoke` operator, `steps` and `dependencies` which i find inconvenient to having it to import all the time. also it makes the usage less self-explaining. maybe it's a good common ground to have some core features and others be considered extensions (as that's what extension features are here for)?

xa17d replied:
> maybe it's a good common ground to have some core features and others (custom ones) be considered extensions...

Valid approach. It's just hard to decide where to draw the line then.
I think it's fine for now. I just wanted to point it out so it's not forgotten.

eder7 commented:
> But thinking about it again, just calling `job.cancel()` at the end of a test consumes a similar amount of brain-power as deciding between `coroutineScope` and `autoCleanupCoroutineScope` or configuring an auto-cleanup.

Just one thing: `autoCleanupCoroutineScope` is a configuration you can put at tests and steps, not an extension with different `CoroutineScope` implementations. There might be a confusion...
Does this change anything for you?

> So, I'd opt for keeping it simple and just don't provide an auto-cleanup option.

Nonetheless, valid point. I also investigated that there is no complaining about still active jobs when there is an exception somewhere in the test. So could be removed without any problem. I'll think about it...

xa17d commented on sweetest/src/main/java/com/mysugr/sweetest/framework/base/BaseSteps.kt:
```diff
@@ -6,7 +6,7 @@ import com.mysugr.sweetest.framework.context.TestContext
 import kotlinx.coroutines.CoroutineScope
 import kotlin.coroutines.CoroutineContext
 
-interface Steps : CoroutineScope
```
I know your reasons for removing this, however, it could be a bit weird that in a test one can write:
```
class BlaTest {
  @Test
  fun `...`() = runBlockingTest {
    launch {
      steps.doSomething()
    }
  }
}
```
...`launch` without explicitly using `coroutineScope`.

And in the steps class:

```
class BlaSteps {
  fun doSomething() {
    coroutineScope.launch { 
      // do something more
    }
  }
}
```
... here the `coroutineScope` is required.


It's a bit inconsistent. Just food for thought.

eder7 replied:
hm, possible solutions:

- just offer `coroutineScope` for steps classes
- don't offer it at all and in steps class have an extension `CoroutineScope.doSomething()` that receives the coroutine scope from the test class / `testCoroutine`.

nonetheless, i consider the extension a convenience thing. question is whether we really need to do something about it and it would have a positive effect or just be annoying...

xa17d commented:
> That's what I did now 👍

Oh, then I overlooked it, nice!

> That's true for old code, but I removed the CoroutineScope interface on Steps, so somehow you need to obtain it. So there is legacy code using it.
>
> However, I moved the legacy extensions to a legacy sub-package. I hope that helps to clarify things and also outlines the usage better.

I didn't think of that, very true 👍

> What does that mean in the current case?

That means we should expose the original `TestCoroutineScope`, without dedicated `delayController`, `uncaughtExceptionCaptor`, etc. But I think we're on the same page now :)

> Just one thing: autoCleanupCoroutineScope is a configuration you can put at tests and steps, not an extension with different CoroutineScope implementations. There might be a confusion...
Does this change anything for you?

Not really. The user still has to decide between options and also has to be aware of those options.

eder7 commented:
OK i'll remove it. Another reason: You're setting "auto cancel" on a class-level even though you don't need in each test function. It might be tempting to enable it and forget about the other test functions where it could mask an error. Thanks for your input on the topic! 🙂

eder7 commented:
See #32
